federationsender/queue/destinationqueue.go

```diff
@@ -51,54 +53,56 @@ type destinationQueue struct {
 	running            atomic.Bool                         // is the queue worker running?
 	backingOff         atomic.Bool                         // true if we're backing off
+	overflowed         atomic.Bool                         // the queues exceed maxPDUsInMemory/maxEDUsInMemory, so we should consult the database for more
 	statistics         *statistics.ServerStatistics        // statistics about this remote server
 	transactionIDMutex sync.Mutex                          // protects transactionID
-	transactionID      gomatrixserverlib.TransactionID     // last transaction ID
-	transactionCount   atomic.Int32                        // how many events in this transaction so far
-	notifyPDUs         chan bool                           // interrupts idle wait for PDUs
-	notifyEDUs         chan bool                           // interrupts idle wait for EDUs
+	transactionID      gomatrixserverlib.TransactionID     // last transaction ID if retrying, or "" if last txn was successful
+	notify             chan struct{}                       // interrupts idle wait pending PDUs/EDUs
+	pendingPDUs        []*queuedPDU                        // PDUs waiting to be sent
+	pendingEDUs        []*queuedEDU                        // EDUs waiting to be sent
+	pendingMutex       sync.RWMutex                        // protects pendingPDUs and pendingEDUs
 	interruptBackoff   chan bool                           // interrupts backoff
 }
 
 // Send event adds the event to the pending queue for the destination.
 // If the queue is empty then it starts a background goroutine to
 // start sending events to that destination.
-func (oq *destinationQueue) sendEvent(receipt *shared.Receipt) {
-	// Create a transaction ID. We'll either do this if we don't have
-	// one made up yet, or if we've exceeded the number of maximum
-	// events allowed in a single tranaction. We'll reset the counter
-	// when we do.
-	oq.transactionIDMutex.Lock()
-	if oq.transactionID == "" || oq.transactionCount.Load() >= maxPDUsPerTransaction {
-		now := gomatrixserverlib.AsTimestamp(time.Now())
-		oq.transactionID = gomatrixserverlib.TransactionID(fmt.Sprintf("%d-%d", now, oq.statistics.SuccessCount()))
-		oq.transactionCount.Store(0)
+func (oq *destinationQueue) sendEvent(event *gomatrixserverlib.HeaderedEvent, receipt *shared.Receipt) {
+	if event == nil {
+		log.Errorf("attempt to send nil PDU with destination %q", oq.destination)
+		return
 	}
-	oq.transactionIDMutex.Unlock()
 	// Create a database entry that associates the given PDU NID with
 	// this destination queue. We'll then be able to retrieve the PDU
 	// later.
 	if err := oq.db.AssociatePDUWithDestination(
 		context.TODO(),
-		oq.transactionID, // the current transaction ID
-		oq.destination,   // the destination server name
-		receipt,          // NIDs from federationsender_queue_json table
+		"",             // TODO: remove this, as we don't need to persist the transaction ID
```

Comment:
Well, I'm in a bit of a philosophical dilemma about this one.

On the one hand, persisting the transaction IDs in the database makes being able to pull things out of the database into `pendingPDUs`/`pendingEDUs` feel a lot more complicated, because you would need to be able to reliably pull out the entire transaction rather than just part of it, and if things are being constantly fed into `pendingPDUs`/`pendingEDUs` on a busy homeserver, we may take a very long time to get back around to there being enough buffer space to do it. 

On the other hand, it does mean that we can lose the transaction ID and have to generate a new one to retry an existing event if Dendrite restarts, and that's not ideal either. I don't quite know which is worse.

Reply:
I may just revert back to the old method of committing the transaction IDs and dealing with them in blocks, as I think that's probably closer to what the spec expects.

Reply:
I've wasted all day trying to make this work and I can't, so I'm sticking with the "don't bother persisting the transaction ID and build new transactions after restart" which is ultimately just what Synapse does.